diff -urN debootstrap-1.0.89.orig/Makefile debootstrap-1.0.89/Makefile
--- debootstrap-1.0.89.orig/Makefile	2017-06-29 10:44:31.071601402 -0400
+++ debootstrap-1.0.89/Makefile	2017-06-29 10:44:49.841601950 -0400
@@ -11,8 +11,7 @@
 	mkdir -p $(DESTDIR)/usr/sbin
 
 	cp -a scripts/* $(DSDIR)/scripts/
-	install -o root -g root -m 0644 functions $(DSDIR)/
+	install -m 0644 functions $(DSDIR)/
 
 	sed 's/@VERSION@/$(VERSION)/g' debootstrap >$(DESTDIR)/usr/sbin/debootstrap
-	chown root:root $(DESTDIR)/usr/sbin/debootstrap
 	chmod 0755 $(DESTDIR)/usr/sbin/debootstrap
