diff -urN libdebian-installer-0.120.orig/configure.ac libdebian-installer-0.120/configure.ac
--- libdebian-installer-0.120.orig/configure.ac	2020-03-25 10:59:41.155908919 -0400
+++ libdebian-installer-0.120/configure.ac	2020-03-25 11:00:25.479243547 -0400
@@ -55,14 +55,7 @@
 fi
 
 AC_MSG_CHECKING([Debian CPU name])
-DEB_HOST_ARCH_CPU="`dpkg-architecture -qDEB_HOST_ARCH_CPU 2>/dev/null`"
-# Take account of old dpkg-architecture output.
-if test -z "$DEB_HOST_ARCH_CPU"; then
-	DEB_HOST_ARCH_CPU="`dpkg-architecture -qDEB_HOST_GNU_CPU`"
-	if test "$DEB_HOST_ARCH_CPU" = x86_64; then
-		DEB_HOST_ARCH_CPU=amd64
-	fi
-fi
+DEB_HOST_ARCH_CPU="powerpc"
 AC_MSG_RESULT([$DEB_HOST_ARCH_CPU])
 
 AC_MSG_CHECKING([Debian system name])
