diff -urN libdebian-installer-0.120.orig/configure.ac libdebian-installer-0.120/configure.ac
--- libdebian-installer-0.120.orig/configure.ac	2020-03-25 10:57:11.209237877 -0400
+++ libdebian-installer-0.120/configure.ac	2020-03-25 10:58:32.809240259 -0400
@@ -1,4 +1,4 @@
-AC_INIT([libdebian-installer],m4_esyscmd(dpkg-parsechangelog | perl -ne 'print $1 if m/^Version: (.*)$/;'))
+AC_INIT([libdebian-installer],[0.121])
 AM_INIT_AUTOMAKE([])
 AM_MAINTAINER_MODE
 
@@ -66,14 +66,7 @@
 AC_MSG_RESULT([$DEB_HOST_ARCH_CPU])
 
 AC_MSG_CHECKING([Debian system name])
-DEB_HOST_ARCH_OS="`dpkg-architecture -qDEB_HOST_ARCH_OS 2>/dev/null`"
-# Take account of old dpkg-architecture output.
-if test -z "$DEB_HOST_ARCH_OS"; then
-	DEB_HOST_ARCH_OS="`dpkg-architecture -qDEB_HOST_GNU_SYSTEM | sed 's/-gnu//g'`"
-	if test "$DEB_HOST_ARCH_OS" = gnu; then
-		DEB_HOST_ARCH_OS=hurd
-	fi
-fi
+DEB_HOST_ARCH_OS="linux"
 AC_MSG_RESULT([$DEB_HOST_ARCH_OS])
 
 if test -f "$srcdir/src/system/subarch-$DEB_HOST_ARCH_CPU-$DEB_HOST_ARCH_OS.c"; then
