diff -urN gnupg-2.2.8/dirmngr/dns.h gnupg-2.2.8.new/dirmngr/dns.h
--- gnupg-2.2.8/dirmngr/dns.h	2017-08-28 06:22:54.000000000 -0400
+++ gnupg-2.2.8.new/dirmngr/dns.h	2020-03-24 13:46:13.359964230 -0400
@@ -522,7 +522,7 @@
 
 
 #define dns_rr_i_new(P, ...) \
-	dns_rr_i_init(&dns_quietinit((struct dns_rr_i){ 0, __VA_ARGS__ }), (P))
+	dns_rr_i_init(dns_quietinit(&(struct dns_rr_i){ 0, __VA_ARGS__ }), (P))
 
 struct dns_rr_i {
 	enum dns_section section;
@@ -1055,7 +1055,7 @@
 #define DNS_OPTS_INITIALIZER  { DNS_OPTS_INITIALIZER_ }
 #define DNS_OPTS_INIT(...)    { DNS_OPTS_INITIALIZER_, __VA_ARGS__ }
 
-#define dns_opts(...) (&dns_quietinit((struct dns_options)DNS_OPTS_INIT(__VA_ARGS__)))
+#define dns_opts(...) (dns_quietinit(&(struct dns_options)DNS_OPTS_INIT(__VA_ARGS__)))
 
 struct dns_options {
 	/*
